### PR TITLE
Add tracing macros to addressresolve

### PR DESCRIPTION
**Pull request by andy31415**
- Implements tracing data structures based on addressresolve (show lookup requests, responses received and failures)
- Add tracing macro calls into the address resolve default implementation
- Centralize the "enable tracing from command line args" and use that in chiptool, minmdns examples (even though these will likely not trace much) and address resolve tool example
- Add json-backend tracing for dnssd data.
- Some bug fixes:
   - fixed proper arg initialization in tracing macros (`{}` instead of `()` for now)
   - make address resolve tool properly shut down resolvers, so we do not segfault on ctrl+c

### CODE REVIEW

File: examples/minimal-mdns/client.cpp [hunk 1]
```diff
@@ -163,6 +169,8 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
                              "        How long to wait for replies\n"
                              "  --multicast-reply\n"
                              "        Do not request unicast replies\n"
+                             "  --trace-to <dest>\n"
+                             "        trace to the given destination.\n"
```
Comment (bzbarsky-apple) [inline examples/minimal-mdns/client.cpp:173]: How does one figure out what destinations are possible/supported?

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp [hunk 2]
```diff
@@ -35,6 +38,8 @@ void NodeLookupHandle::ResetForLookup(System::Clock::Timestamp now, const NodeLo
 
 void NodeLookupHandle::LookupResult(const ResolveResult & result)
 {
+    MATTER_LOG_NODE_DISCOVERED(chip::Tracing::DiscoveryInfoType::kIntermediateResult, &GetRequest().GetPeerId(), &result);
```
Comment (bzbarsky-apple) [inline src/lib/address_resolve/AddressResolve_DefaultImpl.cpp:41]: Why the `chip::` prefixes here and below? We're inside `namespace chip`.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/address_resolve/AddressResolve_DefaultImpl.cpp (hunk 2 above)
Comment (andy31415) [reply src/lib/address_resolve/AddressResolve_DefaultImpl.cpp]: copy and paste. Fixed.

File: examples/minimal-mdns/client.cpp (hunk 1 above)
Comment (andy31415) [reply examples/minimal-mdns/client.cpp]: Added a string constant to help.

Will need to see how this evolves, as I imagine we may want to support logging to files and then the format will become somehow longer - I was thinging of things like `--trace_to log,perfetto,file:/path/to/file` as a readable thing (but we do not yet have that ... so problem for future us)

I imagine I would want some basic target help string here and then have a better README somewhere explaining things.

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**